File: backend/app/api/v1/subscription.py
"""数据资源订阅 API - /api/v1/subscriptions"""
import logging
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession
from typing import Optional

from app.database import get_db
from app.schemas.common import ApiResponse, PaginatedResponse
from app.schemas.subscription import (
    DataSubscriptionCreate, DataSubscriptionUpdate, DataSubscriptionReview, DataSubscriptionResponse,
    DataDeliveryCreate, DataDeliveryResponse,
)
from app.utils.deps import get_current_user, get_pagination_params, PaginationParams
from app.services import subscription_service

logger = logging.getLogger(__name__)

router = APIRouter()


# ==================== 数据订阅 CRUD ====================

@router.post("", response_model=ApiResponse[DataSubscriptionResponse], status_code=201)
async def create_subscription(
    request: DataSubscriptionCreate,
    db: AsyncSession = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """创建数据资源订阅申请"""
    result = await subscription_service.create_subscription(
        db=db,
        subscriber_id=user["user_id"],
        subscriber_org_id=user.get("organization_id", ""),
        catalog_id=request.catalog_id,
        reason=request.reason,
        subscription_config=request.subscription_config,
    )
    return ApiResponse(data=result)


@router.get("", response_model=ApiResponse[PaginatedResponse[DataSubscriptionResponse]])
async def list_subscriptions(
    pagination: PaginationParams = Depends(get_pagination_params),
    status: Optional[str] = Query(None, description="状态筛选"),
    catalog_id: Optional[str] = Query(None, description="目录 ID"),
    db: AsyncSession = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """列出数据订阅"""
    result = await subscription_service.list_subscriptions(
        db, pagination, status, catalog_id, user.get("user_id"), user.get("organization_id")
    )
    return ApiResponse(data=result)


@router.get("/{subscription_id}", response_model=ApiResponse[DataSubscriptionResponse])
async def get_subscription(
    subscription_id: str,
    db: AsyncSession = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """获取订阅详情"""
    result = await subscription_service.get_subscription(db, subscription_id)
    return ApiResponse(data=result)


@router.put("/{subscription_id}", response_model=ApiResponse[DataSubscriptionResponse])
async def update_subscription(
    subscription_id: str,
    request: DataSubscriptionUpdate,
    db: AsyncSession = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """更新订阅"""
    result = await subscription_service.update_subscription(
        db, subscription_id,
        reason=request.reason,
        subscription_config=request.subscription_config,
        status=request.status,
    )
    return ApiResponse(data=result)


@router.post("/{subscription_id}/review", response_model=ApiResponse)
async def review_subscription(
    subscription_id: str,
    request: DataSubscriptionReview,
    db: AsyncSession = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """审核订阅申请"""
    result = await subscription_service.review_subscription(
        db=db,
        subscription_id=subscription_id,
        reviewer_id=user["user_id"],
        status=request.status,
        expires_at=request.expires_at,
    )
    return ApiResponse(data=result)


@router.post("/{subscription_id}/cancel", response_model=ApiResponse)
async def cancel_subscription(
    subscription_id: str,
    db: AsyncSession = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """取消订阅"""
    await subscription_service.cancel_subscription(db, subscription_id, user["user_id"])
    return ApiResponse(message="订阅已取消")


# ==================== 数据交付 ====================

@router.post("/{subscription_id}/deliveries", response_model=ApiResponse[DataDeliveryResponse], status_code=201)
async def create_delivery(
    subscription_id: str,
    request: DataDeliveryCreate,
    db: AsyncSession = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """创建数据交付"""
    result = await subscription_service.create_delivery(
        db=db,
        subscription_id=subscription_id,
        delivery_type=request.delivery_type,
        delivery_config=request.delivery_config,
    )
    return ApiResponse(data=result)


@router.get("/{subscription_id}/deliveries", response_model=ApiResponse[PaginatedResponse[DataDeliveryResponse]])
async def list_deliveries(
    subscription_id: str,
    pagination: PaginationParams = Depends(get_pagination_params),
    db: AsyncSession = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """列出交付记录"""
    result = await subscription_service.list_deliveries(db, pagination, subscription_id)
    return ApiResponse(data=result)


@router.post("/{subscription_id}/deliveries/{delivery_id}/download", response_model=ApiResponse)
async def record_download(
    subscription_id: str,
    delivery_id: str,
    db: AsyncSession = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """记录下载"""
    result = await subscription_service.record_download(db, delivery_id)
    return ApiResponse(data=result)


@router.post("/{subscription_id}/deliveries/{delivery_id}/revoke", response_model=ApiResponse)
async def revoke_delivery(
    subscription_id: str,
    delivery_id: str,
    db: AsyncSession = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """撤销交付"""
    await subscription_service.revoke_delivery(db, delivery_id)
    return ApiResponse(message="交付已撤销")

File: frontend/src/api/subscription.ts
/**
 * 数据资源订阅 API
 * /api/v1/subscriptions - 订阅 CRUD、审核、取消、数据交付管理
 */
import request from './request';
import type {
  ApiResponse, PaginatedResponse, PaginatedRequest,
  DataSubscription, DataDelivery,
} from '@/types/api';

// ==================== 订阅 CRUD ====================

export function createDataSubscription(data: {
  catalog_id: string;
  reason?: string;
  subscription_config?: Record<string, unknown>;
}) {
  return request.post<any, ApiResponse<DataSubscription>>('/subscriptions', data);
}

export function listDataSubscriptions(params?: PaginatedRequest & {
  status?: string;
  catalog_id?: string;
}) {
  return request.get<any, ApiResponse<PaginatedResponse<DataSubscription>>>('/subscriptions', { params });
}

export function getDataSubscription(id: string) {
  return request.get<any, ApiResponse<DataSubscription>>(`/subscriptions/${id}`);
}

export function updateDataSubscription(id: string, data: {
  reason?: string;
  subscription_config?: Record<string, unknown>;
  status?: string;
}) {
  return request.put<any, ApiResponse<DataSubscription>>(`/subscriptions/${id}`, data);
}

export function reviewDataSubscription(id: string, data: {
  status: string;
  expires_at?: string;
}) {
  return request.post<any, ApiResponse<null>>(`/subscriptions/${id}/review`, data);
}

export function cancelDataSubscription(id: string) {
  return request.post<any, ApiResponse<null>>(`/subscriptions/${id}/cancel`);
}

// ==================== 数据交付 ====================

export function createDelivery(subscriptionId: string, data: {
  delivery_type: string;
  delivery_config: Record<string, unknown>;
}) {
  return request.post<any, ApiResponse<DataDelivery>>(`/subscriptions/${subscriptionId}/deliveries`, data);
}

export function listDeliveries(subscriptionId: string, params?: PaginatedRequest) {
  return request.get<any, ApiResponse<PaginatedResponse<DataDelivery>>>(`/subscriptions/${subscriptionId}/deliveries`, { params });
}

export function recordDownload(subscriptionId: string, deliveryId: string) {
  return request.post<any, ApiResponse<null>>(`/subscriptions/${subscriptionId}/deliveries/${deliveryId}/download`);
}

export function revokeDelivery(subscriptionId: string, deliveryId: string) {
  return request.post<any, ApiResponse<null>>(`/subscriptions/${subscriptionId}/deliveries/${deliveryId}/revoke`);
}
